scipy/fft/tests/test_czt.py

```diff
@@ -0,0 +1,215 @@
+    check_zoomfft(x)
+
+
+def test_large_prime_lengths():
+    np.random.seed(0)  # Deterministic randomness
+    for N in (101, 1009, 10007):
+        x = np.random.rand(N)
+        y = fft(x)
+        y1 = czt(x)
+        assert_allclose(y, y1, rtol=1e-12)
+
+
+@pytest.mark.slow
+def test_czt_vs_fft():
+    np.random.seed(123)
+    random_lengths = np.random.exponential(100000, size=10).astype('int')
+    for n in random_lengths:
+        a = np.random.randn(n)
+        assert_allclose(czt(a), fft(a), rtol=1e-11)
+
+
+def test_empty_input():
+    with pytest.raises(ValueError, match='Invalid number of CZT'):
+        czt([])
+    with pytest.raises(ValueError, match='Invalid number of CZT'):
+        zoomfft([], 0.5)
+
+
+def test_0_rank_input():
+    with pytest.raises(IndexError, match='tuple index out of range'):
+        czt(5)
+    with pytest.raises(IndexError, match='tuple index out of range'):
+        zoomfft(5, 0.5)
+
+
+def test_czt_math():
+    for impulse in ([0, 0, 1],
+                    [0, 0, 1, 0, 0],
+                    np.concatenate(([0, 0, 1], np.zeros(100)))):
+        for m in (1, 3, 5, 8, 101, 1021):
+            for a in (1, 2, 0.5, 1.1):
+                for w in (None, 0.7+0.7j):
+                    # z-transform of an impulse is 1 everywhere
+                    assert_allclose(czt(impulse[2:], m=m, a=a),
+                                    np.ones(m), rtol=1e-14)
+
+                    # z-transform of a delayed impulse is z**-1
+                    assert_allclose(czt(impulse[1:], m=m, a=a),
+                                    czt_points(m=m, a=a)**-1, rtol=1e-14)
+
+                    # z-transform of a 2-delayed impulse is z**-2
+                    assert_allclose(czt(impulse, m=m, a=a),
+                                    czt_points(m=m, a=a)**-2, rtol=1e-14)
+
+
+def test_int_args():
+    # Integer argument `a` was producing all 0s
+    assert_allclose(abs(czt([0, 1], m=10, a=2)), 0.5*np.ones(10), rtol=1e-15)
+    assert_allclose(czt_points(11, w=2), 1/(2**np.arange(11)), rtol=1e-30)
+
+
+def test_czt_points():
+    for N in (1, 2, 3, 8, 11, 100, 101, 10007):
+        assert_allclose(czt_points(N), np.exp(2j*np.pi*np.arange(N)/N),
+                        rtol=1e-30)
+
+    assert_allclose(czt_points(7, w=1), np.ones(7), rtol=1e-30)
+    assert_allclose(czt_points(11, w=2.), 1/(2**np.arange(11)), rtol=1e-30)
+
+    func = CZT(12, m=11, w=2., a=1)
+    assert_allclose(func.points(), 1/(2**np.arange(11)), rtol=1e-30)
+
+
+@pytest.mark.parametrize('m', [0, -11, 5.5])
+def test_czt_points_errors(m):
+    # Invalid number of points
+    with pytest.raises(ValueError, match='Invalid number of CZT'):
+        czt_points(m)
+
+
+@pytest.mark.parametrize('myfunc', [CZT(100), ZoomFFT(100, 0.2)])
+def test_invalid_size(myfunc):
+    # Data size doesn't match function's expected size
+    with pytest.raises(ValueError, match='CZT defined for'):
+        myfunc(np.arange(5))
```

Comment:
Maybe something like:
```suggestion
@pytest.mark.parametrize('cls, args', [(CZT, (100,)), (ZoomFFT, (100, 0.2))])
def test_invalid_size(cls, args):
    # Data size doesn't match function's expected size
    myfunc = cls(*args)
    with pytest.raises(ValueError, match='CZT defined for'):
        myfunc(np.arange(5))
```
This is better because the `parametrize` is run at test collection time, so it's better to delay instantiation until the given test is actually run.

Reply:
Why is it better to delay it?

Reply:
Avoid overhead of actual instantiation until within the test run itself. Even if it's negligible here it's a better pattern to follow and universalize.

Also there are side benefits like a stable `__repr__` of the parametrization. If the class doesn't have a custom `__repr__` it will change from run to run because it will by default contain the `id` location in memory.

Reply:
(Imagine there is a class with a 1 s instantiation delay. If you have this in the `parametrize` step, *every* call to `pytest` that includes that file will have this delay, even if for example you do `pytest scipy/signal -k test_firwin` to subselect firwin tests, because test subselection takes place after test collection.)